### ISSUE TITLE
Uncaught exception if client disconnects during async password validation

### ISSUE THREAD

**Issue by pocek** [ronf/asyncssh#84]
asyncssh 1.9.0
Python 3.5.2+ (Ubuntu 16.10)

If `SSHServer.validate_password()` is a coroutine (legal API use) and the client disconnects during its execution, there is uncaught exception being logged:

```
$ ./server.py
running
connection made
validating password... (now close the client)
connection lost
DEBUG:asyncssh:Uncaught exception
Traceback (most recent call last):
  File "/home/kjk/.local/lib/python3.5/site-packages/asyncssh/connection.py", line 374, in _run_task
    yield from coro
  File "/home/kjk/.local/lib/python3.5/site-packages/asyncssh/auth.py", line 749, in _validate_response
    responses)
  File "/home/kjk/.local/lib/python3.5/site-packages/asyncssh/connection.py", line 3441, in validate_kbdint_response
    result = yield from result
  File "/usr/lib/python3.5/asyncio/coroutines.py", line 124, in throw
    return self.gen.throw(exc)
  File "./server.py", line 17, in validate_password
    await asyncio.sleep(60)
  File "/usr/lib/python3.5/asyncio/tasks.py", line 517, in sleep
    return (yield from future)
  File "/usr/lib/python3.5/asyncio/futures.py", line 379, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/lib/python3.5/asyncio/tasks.py", line 297, in _wakeup
    future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 284, in result
    raise CancelledError
concurrent.futures._base.CancelledError
```

Unfortunately catching `asyncio.CancelledError` in `SSHServer.validate_password()` and returning something meaningful doesn't do much good:
```
$ ./server.py
running
connection made
validating password... (now close the client)
connection lost
cancelled
DEBUG:asyncssh:Uncaught exception
Traceback (most recent call last):
  File "./server.py", line 17, in validate_password
    await asyncio.sleep(60)
  File "/usr/lib/python3.5/asyncio/tasks.py", line 517, in sleep
    return (yield from future)
  File "/usr/lib/python3.5/asyncio/futures.py", line 379, in __iter__
    yield self  # This tells Task to wait for completion.
  File "/usr/lib/python3.5/asyncio/tasks.py", line 297, in _wakeup
    future.result()
  File "/usr/lib/python3.5/asyncio/futures.py", line 284, in result
    raise CancelledError
concurrent.futures._base.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/kjk/.local/lib/python3.5/site-packages/asyncssh/connection.py", line 374, in _run_task
    yield from coro
  File "/home/kjk/.local/lib/python3.5/site-packages/asyncssh/auth.py", line 750, in _validate_response
    self._send_challenge(next_challenge)
  File "/home/kjk/.local/lib/python3.5/site-packages/asyncssh/auth.py", line 741, in _send_challenge
    self.send_failure()
  File "/home/kjk/.local/lib/python3.5/site-packages/asyncssh/auth.py", line 467, in send_failure
    self._conn.send_userauth_failure(partial_success)
  File "/home/kjk/.local/lib/python3.5/site-packages/asyncssh/connection.py", line 1001, in send_userauth_failure
    NameList(get_server_auth_methods(self)),
  File "/home/kjk/.local/lib/python3.5/site-packages/asyncssh/auth.py", line 847, in get_server_auth_methods
    if _server_auth_handlers[method].supported(conn):
  File "/home/kjk/.local/lib/python3.5/site-packages/asyncssh/auth.py", line 665, in supported
    return conn.public_key_auth_supported()
  File "/home/kjk/.local/lib/python3.5/site-packages/asyncssh/connection.py", line 3344, in public_key_auth_supported
    self._owner.public_key_auth_supported())
AttributeError: 'NoneType' object has no attribute 'public_key_auth_supported'
```

Simple program used to reproduce this:

```python
#!/usr/bin/python3

import asyncio
import asyncssh
import logging

class MySSHServer(asyncssh.SSHServer):
    def connection_made(self, conn):
        print('connection made')
    def connection_lost(self, exc):
        print('connection lost')
    def password_auth_supported(self):
        return True
    async def validate_password(self, username, password):
        print('validating password... (now close the client)')
        try:
            await asyncio.sleep(60)
        except asyncio.CancelledError:
            print('cancelled')
            return False
        return False

logging.basicConfig(level=logging.DEBUG)
logging.getLogger('asyncio').setLevel(logging.WARNING)
loop = asyncio.get_event_loop()
loop.set_debug(True)
loop.slow_callback_duration = 200
loop.run_until_complete(asyncssh.create_server(MySSHServer,
    '127.0.0.1', 8022, server_host_keys=['ssh_host_key']))
print('running')
loop.run_forever()
```

with combination with `ssh -p 8022 localhost` and hitting Ctrl+C after the password is entered.

**Comment by ronf** [comment]
Thank you for the bug report! It looks like the call to send the userauth failure is happening after the cleanup function is called on the connection. The following change should fix this:

```patch
--- a/asyncssh/connection.py
+++ b/asyncssh/connection.py
@@ -996,6 +996,9 @@ class SSHConnection(SSHPacketHandler):
     def send_userauth_failure(self, partial_success):
         """Send a user authentication failure response"""
 
+        if not self._owner:
+            return
+
         self._auth = None
         self.send_packet(Byte(MSG_USERAUTH_FAILURE),
                          NameList(get_server_auth_methods(self)),
```

I'll be checking this fix into the "develop" branch shortly and I'll include it in the next release.

There may be other races of this sort now that some of the callbacks in SSHConnection are allowed to be coroutines. A fix similar to the one above to simply exit without any further processing when owner is no longer set should address this, though. Please let me know if you run into this in any other places.

**Comment by ronf** [comment]
Actually, thinking about this a bit more, there's a better fix. I'm explicitly cancelling the auth future in cleanup, so I really should be handling the CancelledError as well in my task wrapper. This actually avoids the need to worry about a task continuing to run after cleanup is called. Here's the new fix:

```patch
--- a/asyncssh/connection.py
+++ b/asyncssh/connection.py
@@ -372,6 +372,8 @@ class SSHConnection(SSHPacketHandler):
         # pylint: disable=broad-except
         try:
             yield from coro
+        except CancelledError:
+            pass
         except DisconnectError as exc:
             self._send_disconnect(exc.code, exc.reason, exc.lang)
             self._force_close(exc)
```

**Comment by ronf** [comment]
With the new fix, you also don't need to worry about catching the CancelledError yourself in the callback. Changing your test callback to the following works fine now:

```python
    async def validate_password(self, username, password):
        print('validating password... (now close the client)')
        await asyncio.sleep(60)
        return False
```

**Comment by pocek** [comment]
> The version of your test which tries to catch CancelledError fails without the self._owner check, but I'm thinking that trying to catch that error and return rather than re-raising it is probably not something it makes sense to support.

I've just tried ce32a28 and it's indeed fine that way. I think the need to re-raise - in case someone actually wants to handle cancellation - is worth documenting though.

Thank you for the fix (and asyncssh as a whole - it's great!).

**Comment by ronf** [comment]
Thanks for the confirmation, pocek!

Yes, re-raising the CancelledError should work, if there's a need to do local cleanup when the connection unexpectedly closes.

I'll see if I can find a good place to document this.

**Comment by ronf** [comment]
Ok - I've added a note about this to the top-level of the SSHClient and SSHServer classes.

**Comment by ronf** [comment]
This fix is now available in AsyncSSH version 1.10.0.
